test/common/common/url_template_matcher_test.cc

```diff
@@ -0,0 +1,65 @@
+#include "envoy/common/exception.h"
+
+#include "common/common/url_template_matcher.h"
+
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Matchers {
+namespace {
+
+TEST(UrlTemplateMatcherTest, OnlyLiteral) {
+  UrlTemplateMatcher m("/foo/bar");
+  EXPECT_TRUE(m.match("/foo/bar"));
+  EXPECT_FALSE(m.match("/bar"));
+}
+
+TEST(UrlTemplateMatcherTest, WildCard) {
+  // * match 1 segment.
+  UrlTemplateMatcher m("/foo/*");
+  EXPECT_TRUE(m.match("/foo/bar"));
+  EXPECT_FALSE(m.match("/foo"));
+  EXPECT_FALSE(m.match("/foo/zoo/bar"));
+  EXPECT_FALSE(m.match("/bar"));
+}
+
+TEST(UrlTemplateMatcherTest, WildCardWithVariable) {
+  // {x} is the same as *: match 1 segment.
+  UrlTemplateMatcher m("/foo/{x}");
+  EXPECT_TRUE(m.match("/foo/bar"));
+  EXPECT_FALSE(m.match("/foo"));
+  EXPECT_FALSE(m.match("/foo/zoo/bar"));
+  EXPECT_FALSE(m.match("/bar"));
+}
+
+TEST(UrlTemplateMatcherTest, DoubleWildCard) {
+  // ** match zero or more segments
+  UrlTemplateMatcher m("/foo/**");
+  EXPECT_TRUE(m.match("/foo"));
+  EXPECT_TRUE(m.match("/foo/bar"));
+  EXPECT_TRUE(m.match("/foo/zoo/bar"));
+  EXPECT_FALSE(m.match("/bar"));
+}
+
+TEST(UrlTemplateMatcherTest, DoubleWildCardWithVariable) {
+  // ** match zero or more segments
+  UrlTemplateMatcher m("/foo/**/tail");
+  EXPECT_TRUE(m.match("/foo/bar/tail"));
+  EXPECT_TRUE(m.match("/foo/zoo/bar/tail"));
+
+  EXPECT_FALSE(m.match("/foo/other"));
+  EXPECT_FALSE(m.match("/foo/bar/other"));
+  EXPECT_FALSE(m.match("/foo/zoo/bar/other"));
+
+  EXPECT_FALSE(m.match("/bar/tail"));
+}
+
```

Comment:
It's possible to escape syntax characters with [percent encoding](https://github.com/googleapis/googleapis/blob/f829b1334cce86aa3738f3c0698d814b56664445/google/api/http.proto#L242):
> // The syntax `LITERAL` matches literal text in the URL path. If the `LITERAL`
> // contains any reserved character, such characters should be percent-encoded
> // before the matching.

It probably makes sense to have tests that make sure that behavior works the way we expect.  For example:
 * Template `/foo/%2A/*` would match `/foo/*/bar` and `/foo/*/foo` but not `/foo/blah/foo` 
 * Template `/foo/%2A%2A/**` would match `/foo/**/bar`, `/foo/**/foo`, and `/foo/**` but not `/foo/blah/foo`.

Reply:
Yes,  both url_template and path should escape these characters.  OK, will add tests for it

Reply:
I still think we could benefit from a test where the `%2A` are in the template and there are `*` in the URL.

```
TEST(UrlTemplateMatcherTest, EscapedStarInTemplate) {
  UrlTemplateMatcher m("/foo/%2A/*");
  EXPECT_TRUE(m.match("/foo/*/bar"));
  EXPECT_TRUE(m.match("/foo/*/foo"));
  EXPECT_FALSE(m.match("/foo/blah/foo"));
}

TEST(UrlTemplateMatcherTest, EscapedStarStarInTemplate) {
  UrlTemplateMatcher m("/foo/%2A%2A/**");
  EXPECT_TRUE(m.match("/foo/**/bar"));
  EXPECT_TRUE(m.match("/foo/**/foo"));
  EXPECT_TRUE(m.match("/foo/**"));
  EXPECT_FALSE(m.match("/foo/blah/foo"));
}
```